v1/src/main/java/com/google/cloud/teleport/datadog/DatadogEventWriter.java

```diff
@@ -0,0 +1,463 @@
+      "Invalid url format. Url format should match PROTOCOL://HOST[:PORT], where PORT is optional. "
+          + "Supported Protocols are http and https. eg: http://hostname:8088";
+
+  @StateId(BUFFER_STATE_NAME)
+  private final StateSpec<BagState<DatadogEvent>> buffer = StateSpecs.bag();
+
+  @StateId(COUNT_STATE_NAME)
+  private final StateSpec<ValueState<Long>> count = StateSpecs.value();
+
+  @TimerId(TIME_ID_NAME)
+  private final TimerSpec expirySpec = TimerSpecs.timer(TimeDomain.EVENT_TIME);
+
+  private Integer batchCount;
+  private HttpEventPublisher publisher;
+
+  private static final Gson GSON =
+      new GsonBuilder().setFieldNamingStrategy(f -> f.getName().toLowerCase()).create();
+
+  public static Builder newBuilder() {
+    return new AutoValue_DatadogEventWriter.Builder();
+  }
+
+  @Nullable
+  abstract ValueProvider<String> url();
+
+  @Nullable
+  abstract ValueProvider<String> apiKey();
+
+  @Nullable
+  abstract ValueProvider<Integer> inputBatchCount();
+
+  @Setup
+  public void setup() {
+
+    checkArgument(url().isAccessible(), "url is required for writing events.");
+    checkArgument(isValidUrlFormat(url().get()), INVALID_URL_FORMAT_MESSAGE);
+    checkArgument(apiKey().isAccessible(), "API Key is required for writing events.");
+
+    // Either user supplied or default batchCount.
+    if (batchCount == null) {
+
+      if (inputBatchCount() != null) {
+        batchCount = inputBatchCount().get();
+      }
+
+      batchCount = MoreObjects.firstNonNull(batchCount, DEFAULT_BATCH_COUNT);
+      LOG.info("Batch count set to: {}", batchCount);
+    }
+    checkArgument(
+        batchCount <= MAX_BATCH_COUNT,
+        "batchCount must be less than or equal to %s",
+        MAX_BATCH_COUNT);
+
+    try {
+      HttpEventPublisher.Builder builder =
+          HttpEventPublisher.newBuilder().withUrl(url().get()).withApiKey(apiKey().get());
+
+      publisher = builder.build();
+      LOG.info("Successfully created HttpEventPublisher");
+
+    } catch (NoSuchAlgorithmException | KeyManagementException | IOException e) {
+      LOG.error("Error creating HttpEventPublisher: {}", e.getMessage());
+      throw new RuntimeException(e);
+    }
+  }
+
+  @ProcessElement
+  public void processElement(
+      @Element KV<Integer, DatadogEvent> input,
+      OutputReceiver<DatadogWriteError> receiver,
+      BoundedWindow window,
+      @StateId(BUFFER_STATE_NAME) BagState<DatadogEvent> bufferState,
+      @StateId(COUNT_STATE_NAME) ValueState<Long> countState,
+      @TimerId(TIME_ID_NAME) Timer timer)
+      throws IOException {
+
+    Long count = MoreObjects.<Long>firstNonNull(countState.read(), 0L);
+    DatadogEvent event = input.getValue();
+    INPUT_COUNTER.inc();
+    bufferState.add(event);
+    count += 1;
+    countState.write(count);
+    timer.offset(Duration.standardSeconds(DEFAULT_FLUSH_DELAY)).setRelative();
+
+    if (count >= batchCount) {
+      LOG.debug("Flushing batch of {} events", count);
```

Comment:
https://cloud.google.com/dataflow/docs/guides/logging#SettingLevels should be the right place. it's a bit tricky for Templates, though. a lot of parameters are defined in build time and not runtime (unless ValueProviders).

Curious if there's any special reason to create a v1 (Classic) instead of v2 (Flex)?

Reply:
Yes, I did see this documentation! However, it seems to me that documentation is for a Dataflow developer to modify log level via pipeline options in template source code, not for a Dataflow user to modify the log level when running a job (i.e. `--defaultSdkHarnessLogLevel` is not accepted by `gcloud dataflow jobs run`). So AFAICT, there's no way for a Dataflow user to easily change (i.e. via some parameter) the log level for a Dataflow job. That's why I was asking if what's here is even useful?

> it's a bit tricky for Templates, though. a lot of parameters are defined in build time and not runtime  (unless ValueProviders)

Understood. I see in the staged template files some hardcoded `sdkPipelineOptions`, so yea, build time. 👍 

> Curious if there's any special reason to create a v1 (Classic) instead of v2 (Flex)

I think we've touched on this before, but to reiterate: There was an existing classic template that very much fit our use case, so we decided to use it as a base rather than work from scratch. From our understanding, Flex would be helpful if we needed to use a sink or source that didn't support `ValueProvider`s, but that was not the case for us.